crates/services/txpool/src/transaction_selector.rs

```diff
@@ -6,27 +6,43 @@ use fuel_core_types::{
 // transaction selection could use a plugin based approach in the
 // future for block producers to customize block building (e.g. alternative priorities besides gas fees)
 
-// Expects sorted by gas price transactions, highest first
+// Expects sorted by tip, highest first
 pub fn select_transactions(
     includable_txs: impl Iterator<Item = ArcPoolTx>,
     max_gas: u64,
-) -> Vec<ArcPoolTx> {
-    // Select all txs that fit into the block, preferring ones with higher gas price.
-    //
+    block_transaction_size_limit: u64,
+) -> impl Iterator<Item = ArcPoolTx> {
     // Future improvements to this algorithm may take into account the parallel nature of
     // transactions to maximize throughput.
-    let mut used_block_space: Word = 0;
+
+    let mut used_block_gas_space: Word = 0;
+    let mut used_block_size_space: Word = 0;
+
     // The type of the index for the transaction is `u16`, so we need to
     // limit it to `MAX` value minus 1(because of the `Mint` transaction).
     let takes_txs = u16::MAX - 1;
 
-    // Pick as many transactions as we can fit into the block (greedy)
+    // Pick as many transactions as we can fit into the block (greedy),
+    // respecting both gas and size limit
     includable_txs
-        .filter(|tx| {
-            let tx_block_space = tx.max_gas();
-            if let Some(new_used_space) = used_block_space.checked_add(tx_block_space) {
-                if new_used_space <= max_gas {
-                    used_block_space = new_used_space;
+        .filter(move |tx| {
```

Comment:
It is not part of this PR, but filtering *each* entry of the transaction pool is too expensive. We should rework it and stop iteration when we filled our limits.

Reply:
I created a follow-up issue https://github.com/FuelLabs/fuel-core/issues/2235